### PR TITLE
Resume Storage Market Client Data Transfer 

### PR DESCRIPTION
**Pull request by aarshkshah1992**
For https://github.com/filecoin-project/lotus/issues/3417.

### CODE REVIEW

File: storagemarket/impl/dtutils/dtutils.go [hunk 1]
```diff
@@ -82,6 +87,16 @@ func ClientDataTransferSubscriber(deals EventReceiver) datatransfer.Subscriber {
 
 		// data transfer events for progress do not affect deal state
 		switch event.Code {
+		case datatransfer.Restart:
+			err := deals.Send(voucher.Proposal, storagemarket.ClientEventDataTransferRestarted)
+			if err != nil {
+				log.Errorf("processing dt event: %w", err)
+			}
+		case datatransfer.Accept:
+			err := deals.Send(voucher.Proposal, storagemarket.ClientEventDataTransferInitiated, channelState.ChannelID())
```
Comment (aarshkshah1992) [inline storagemarket/impl/dtutils/dtutils.go:96]: @hannahhoward  

The reason I have moved this here is because if we restart a failed transfer on the client side, we need to be sure that the provider has already created the channel which is what `Accept` will tell us.

Let me know what you think.

File: storagemarket/integration_test.go [hunk 2]
```diff
@@ -277,65 +281,103 @@ func TestMakeDealNonBlocking(t *testing.T) {
...
 	pd := providerDeals[0]
 	assert.Equal(t, result.ProposalCid, pd.ProposalCid)
-	shared_testutil.AssertDealState(t, storagemarket.StorageDealProviderFunding, pd.State)
+	require.Eventually(t, func() bool {
+		return pd.State == storagemarket.StorageDealProviderFunding
+	}, 1*time.Second, 100*time.Millisecond, "actual deal status is %s", storagemarket.DealStates[pd.State])
 }
 
+// FIXME Gets hung sometimes
 func TestRestartClient(t *testing.T) {
```
Comment (aarshkshah1992) [inline storagemarket/integration_test.go:290]: @hannahhoward Some test cases here can fail in hairy edge cases. I've documented those edge cases for each test as comments on the test case.

Please let me know what you think about handling them. If they seem rare enough, we can just let it be and avoid the complexity.

File: storagemarket/integration_test.go [hunk 3]
```diff
@@ -277,65 +281,104 @@ func TestMakeDealNonBlocking(t *testing.T) {
...
-		"ClientEventFundsEnsured": {
-			stopAtEvent: storagemarket.ClientEventFundsEnsured,
-		},
-		"ClientEventInitiateDataTransfer": {
-			stopAtEvent: storagemarket.ClientEventInitiateDataTransfer,
-		},
+
 		"ClientEventDataTransferInitiated": {
-			stopAtEvent: storagemarket.ClientEventDataTransferInitiated,
+			// This test can fail if client crashes without seeing a Provider DT complete
```
Comment (aarshkshah1992) [inline storagemarket/integration_test.go:300]: @hannahhoward This is a very tricky one. If the client crashes before processing a `Complete Event` but the Provider has already completed it's FSM, the `Restart` will fail later as the channel will NOT exist in the Provider's FSM.

Comment (aarshkshah1992) [reply storagemarket/integration_test.go:300]: Do we want to fix this ?

File: storagemarket/integration_test.go [hunk 4]
```diff
@@ -277,65 +281,104 @@ func TestMakeDealNonBlocking(t *testing.T) {
...
+			stopAtEvent:         storagemarket.ClientEventDataTransferInitiated,
+			expectedClientState: storagemarket.StorageDealTransferring,
+			clientDelay:         noOpDelay,
+			providerDelay:       noOpDelay,
 		},
+
 		"ClientEventDataTransferComplete": {
-			stopAtEvent: storagemarket.ClientEventDataTransferComplete,
+			// fails if client is in StorageDealCheckForAcceptance but
+			// provider is in StorageDealTransferring after restarts
```
Comment (aarshkshah1992) [inline storagemarket/integration_test.go:309]: This could be solved by restarting correctly on the Provider.

File: storagemarket/integration_test.go [hunk 5]
```diff
@@ -277,65 +281,104 @@ func TestMakeDealNonBlocking(t *testing.T) {
...
+			expectedClientState: storagemarket.StorageDealCheckForAcceptance,
 		},
-		"ClientEventDealAccepted": {
-			stopAtEvent: storagemarket.ClientEventDealAccepted,
+
+		"ClientEventFundsEnsured": {
+			//
+			//Edge case : Provider begins the state machine on recieving a deal stream request
+			//client crashes -> restarts -> sends deal stream again -> state machine fails
+			stopAtEvent:         storagemarket.ClientEventFundsEnsured,
```
Comment (aarshkshah1992) [inline storagemarket/integration_test.go:321]: TODO: File a fleshed out issue.

**Comment by aarshkshah1992** [comment]
@hannahhoward Have documented the failing tests in https://github.com/filecoin-project/lotus/issues/3966.

**Review by hannahhoward — APPROVED** [review]
This is great @aarshkshah1992 merging to your branch
